aten/src/ATen/native/sparse/cuda/SparseCUDATensor.cu

```diff
@@ -19,11 +19,57 @@
 #include <thrust/transform.h>
 #include <thrust/unique.h>
 #include <thrust/system/cuda/execution_policy.h>
+#include <thrust/binary_search.h>
 #include <c10/macros/Macros.h>
 
 namespace at { namespace native {
 
 using namespace at::sparse;
+using at::cuda::detail::TensorInfo;
+using at::cuda::detail::getTensorInfo;
+
+namespace {
+
+template <typename scalar_t>
+#ifdef __HIP_PLATFORM_HCC__
+C10_LAUNCH_BOUNDS_1(512)
+#endif
+__global__ void _sparse_mask_copy_kernel(
+  int64_t total_threads,
+  int64_t t_nnz,
+  const TensorInfo<int64_t, int64_t> t_indices_ti,
+  const TensorInfo<int64_t, int64_t> mask_indices_ti,
+  const TensorInfo<int64_t, int64_t> t_indices_pos_ti,
+  const TensorInfo<scalar_t, int64_t> t_values_ti,
+  TensorInfo<scalar_t, int64_t> r_values_ti
+) {
+  const int64_t i = blockIdx.x * blockDim.x + threadIdx.x;
+  if (i >= total_threads) return;
+  const int64_t j = t_indices_pos_ti.data[i];
+
+  bool has_match = false;
+  if (j >= 0 &&  j < t_nnz && t_indices_ti.data[j] == mask_indices_ti.data[i]) {
+    has_match = true;
+  }
+
+  int64_t values_stride0 = r_values_ti.strides[0];
+  int64_t out_start = i * values_stride0;
+  int64_t out_end = (i + 1) * values_stride0;
+  int64_t in_start = j * t_values_ti.strides[0];
+
+  if (has_match) {
```

Comment:
Is this necessary to handle hybrid sparse-dense tensors?

Reply:
This is necessary to copy just the values that has a match. The else part was  removed because the `r_values` are initialized as `Tensor r_values = at::zeros({vsize}, t_values.options());`